File: gfStock/src/main/java/job/gf/simplejms/server/protocol/MsgFrame.java
package job.gf.simplejms.server.protocol;

public class MsgFrame extends Frame{
	public static MsgFrame ZERO_MSGFRAME=new MsgFrame(0);
	
	private int writeLen=0;
	
	public MsgFrame(int totalLen){
		super(totalLen);
	}


	public int getWriteLen() {
		return writeLen;
	}
	
	public int getRestLen() {
		return totalLen-writeLen;
	}
	
	public int writeData(byte[] d,int start){
		if(d==null || d.length==0 || isOk()) return 0;
		
		int len = d.length-start;
		int restLen = getRestLen();
		len = Math.min(len, restLen);
		if(len==0) return 0;
		
		System.arraycopy(d, start, data, writeLen, len);
		writeLen+=len;
		return len;
	}


	public int writeData(byte[] d){
		return writeData(d,0);
	}
	
	public boolean isOk(){
		return writeLen==totalLen;
	}
}

File: gfStockNodeServer/lib/protocol/MsgFrame.js
exports = module.exports=function MsgFrame(stotalLen){
	totalLen=stotalLen;
	data=new Buffer(stotalLen);
	writeLen=0;
	
	//isOk
	this.isOk=function(){
		return writeLen==totalLen;
	}
	
	this.getRestLen=function() {
		return totalLen-writeLen;
	}
	
	this.getWriteLen=function(){
		return writeLen;
	}
	
	this.getTotalLen=function(){
		return totalLen;
	}
	
	this.getData=function(){
		return data;
	}
	
	this.writeData=function(d,start){
		if(d==null || d.length==0 || this.isOk()) return 0;
		
		var len = d.length-start;
		var restLen = this.getRestLen();
		len = Math.min(len, restLen);
		if(len==0) return 0;
		
		for(var i=0;i<len;i++){
			data[i+writeLen]=d[start+i];
		}
		writeLen+=len;
		return len;
	}
}


